# Анализ базы данных крупного сервиса для чтения книг по подписке

### Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Задача — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<b>Цель исследования</b> - провести анализ базы данных крупного сервиса для чтения книг по подписке, на основании которого необходимо сформулировать предложения для нового продукта.

<div>
<b>Описание данных</b><br>

Таблица `books`
    
Содержит данные о книгах:
* `book_id` — идентификатор книги;
* `author_id`— идентификатор автора;  
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица `authors`
    
Содержит данные об авторах:
* `author_id`— идентификатор автора;
* `author` — имя автора.
    
Таблица `publishers`
    
Содержит данные об издательствах:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства.
    
Таблица `ratings`
    
Содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица `reviews`
    
Содержит данные о пользовательских обзорах на книги:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора.
  

### Откройте файлы с данными и изучите общую информацию

In [1]:
!pip install psycopg2-binary

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
#объявим запрос
query = '''
            SELECT *
            FROM books
            LIMIT 10
'''

In [5]:
#исследуем датасет book
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [6]:
#объявим запрос
query = '''
            SELECT *
            FROM authors
            LIMIT 10
'''

In [7]:
#исследуем датасет authors
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [8]:
#объявим запрос
query = '''
            SELECT *
            FROM publishers
            LIMIT 10
'''

In [9]:
#исследуем датасет publishers
pd.io.sql.read_sql(query, con = engine) 

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [10]:
#объявим запрос
query = '''
            SELECT *
            FROM ratings
            LIMIT 10
'''

In [11]:
#исследуем датасет ratings
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [12]:
#объявим запрос
query = '''
            SELECT *
            FROM reviews
            LIMIT 10
'''

In [13]:
#исследуем датасет reviews
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [14]:
#объявим запрос
query = '''
            SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01'
            
'''

In [15]:
#выведем полученный запрос
pd.io.sql.read_sql(query, con = engine)

count
0    819

**Вывод:**

<p></p>

<div style="border:solid green 2px; padding: 20px">
После 1 января 2000 года вышло в каталоге нового сервиса 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [16]:
#объявим запрос
query = '''
            SELECT b.book_id,
                   b.title,
                   COUNT (DISTINCT re.review_id) AS count_review,
                   ROUND(AVG(rating),2) AS avg_rating
            FROM books AS b
            LEFT JOIN reviews AS re ON b.book_id = re.book_id
            LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
            
            GROUP BY b.book_id, b.title
            ORDER BY count_review DESC
            
'''

In [17]:
#выведем полученный запрос
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  count_review  \
0        948                            Twilight (Twilight  #1)             7   
1        963                                Water for Elephants             6   
2        734                                   The Glass Castle             6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
4        695  The Curious Incident of the Dog in the Night-Time             6   
..       ...                                                ...           ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996      808                            The Natural Way to Draw             0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998      221                          Essential Tales and Poems             0   
999      191  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

**Вывод:**

<p></p>

<div style="border:solid green 2px; padding: 20px">
Общее количество книг и брошюр в катологе составляет 1000 штук. Количество обзоров на книгу находится в диапазоне от 0 до 7. То есть в датасете присутствуют книги, на которые нет обзоров. Так, лидер по количеству обзоров книга "Twilight (Twilight #1)", количество обзоров 7, средний рейтинг 3,66. Если обратить внимание на средную оценку, то можно увидеть, что количество обзоров никак не связано со средней оценкой, то есть корреляция отсутствует. Есть книга, например, "The Cat in the Hat and Other Dr. Seuss Favorites", у которой нет обзоров, при этом средний рейтинг = 5. 

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [18]:
#объявим запрос
query = '''
            WITH
            t_1 AS
            (SELECT publisher_id,
                    book_id
            FROM books
            WHERE num_pages > 50)
            SELECT t_1.publisher_id,
                   p.publisher
            FROM t_1
            INNER JOIN publishers AS p ON t_1.publisher_id = p.publisher_id
            GROUP BY t_1.publisher_id, p.publisher 
            ORDER BY COUNT(t_1.book_id) DESC
            LIMIT 1
'''

In [19]:
#выведем полученный запрос
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher
0           212  Penguin Books

**Вывод:**

<p></p>

<div style="border:solid green 2px; padding: 20px">
По условию задачи были исключены из анализа книги, в которых меньше 50 страниц, так как они относятся к брошюрам. Издательство, которое выпустило больше всего книг толще 50 страниц - Penguin Books (британское издательство). 

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [20]:
#объявим запрос
query = '''
            WITH 
            t1 AS
            (SELECT book_id
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(rating_id) >= 50)
            SELECT a.author_id, 
                   a.author, 
                   AVG(r.rating) AS avg_rating
            FROM books AS b

            INNER JOIN t1 ON b.book_id = t1.book_id
            INNER JOIN authors AS a ON b.author_id = a.author_id
            INNER JOIN ratings AS r ON t1.book_id = r.book_id
            GROUP BY a.author_id, a.author
            ORDER BY AVG(r.rating) DESC
            LIMIT 1

'''

In [21]:
#выведем полученный запрос
pd.io.sql.read_sql(query, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

**Вывод:**

<p></p>

<div style="border:solid green 2px; padding: 20px">
По условию задачи были исключены книги из анализа, в которых меньше 50 оценок. Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré. Средняя оценка составила 4,28. Джоан Роулинг самая популярная британская писательница, написавшая серию книг о Гарри Поттере. 

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [22]:
#объявим запрос
query = '''
           SELECT COUNT(DISTINCT rv.text)/ COUNT(DISTINCT rv.username) AS avg_rv_count
           FROM ratings AS r

           INNER JOIN reviews AS rv ON r.username = rv.username
           WHERE r.username in  
                    (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50)

'''

In [23]:
#выведем полученный запрос
pd.io.sql.read_sql(query, con = engine)

avg_rv_count
0            24

**Вывод:**

<p></p>

<div style="border:solid green 2px; padding: 20px">
Пользователи, которые поставили больше 50 оценок, написали в среднем обзоры на 24 книги и брошюры. 

**Общие выводы:**

<p></p>

1. После 1 января 2000 года вышло в каталоге нового сервиса 819 книг.<br><br>

2. Общее количество книг и брошюр в катологе составляет 1000 штук. Количество обзоров на книгу находится в диапазоне от 0 до 7. То есть в датасете присутствуют книги, на которые нет обзоров. Так, лидер по количеству обзоров книга "Twilight (Twilight #1)", количество обзоров 7, средний рейтинг 3,66. Если обратить внимание на средную оценку, то можно увидеть, что количество обзоров никак не связано со средней оценкой, то есть корреляция отсутствует. Есть книга, например, "The Cat in the Hat and Other Dr. Seuss Favorites", у которой нет обзоров, при этом средний рейтинг = 5. <br><br>

3. Издательство, которое выпустило больше всего книг толще 50 страниц - Penguin Books (британское издательство). <br><br>

4. Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré. Средняя оценка составила 4,28. Джоан Роулинг самая популярная британская писательница, написавшая серию книг о Гарри Поттере.  <br><br>

5. Пользователи, которые поставили больше 50 оценок, написали в среднем обзоры на 24 книги и брошюры. 

**Рекомендации**<br>

В качестве предложений для разработки нового продукта можно:
1. Расширить базу данных, добавить новые книги, можно посмотреть книги в разных издательствах и собрать в одну библиотеку разные годы изданий. 
2. Внедрить маркетинговую активность, которая позволит собрать оценки пользователей по всем книгам и брошюрам, имеющимся в базе данных. Это как отзывы, которые мотивируют совершать действие. Например, в онлайн-магазине это покупка, в онлайн-библиотеке будет подписка на прочтение книги. 
3. Увеличить ассортимент книг и жанров самого популярного  британского издательства Penguin Books.
4. Добавить другие серии книг самой популярной британской писательницы Джоан Роулинг.
5. Давать бонусы за оставление оценки по новым вышедшим книгам, размещенным в онлайн-библиотеке. Можно в виде скидки на подписку на месяц. 